### PR TITLE
Categories breadcrumbs optimization for Large Category Tree

### PR DESCRIPTION
**Pull request by AntoInta**
We tested the original solution in a tree with more than 350 categories and result is an Out Of Memory (PHP 7.1, memory_limit = 2G). The proposed solution solves the problem for large category trees (~400ms in the same host).

Every suggestion is really appreciated ;)

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | "develop"
| Type?         | improvement
|Description   | The proposed solution redefine the getCategoriesWithBreadCrumb() method in order to retrieve all the breadcrumbs for every category directly from a query, thanks to the binary search attributes.
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| How to test?  | Is it possible to compare the current  and the proposed solution with a large Category Tree, accessing in the BackOffice in Product Info page.

#### Important guidelines

* Make sure [your local branch is up to date](https://help.github.com/articles/syncing-a-fork/) before commiting your changes!
* Your code MUST respect [our Coding Standards](http://doc.prestashop.com/display/PS16/Coding+Standards) (for code written in PHP, JavaScript, HTML/CSS/Smarty/Twig, SQL)!
* Your commit name MUST respect our [naming convention](http://doc.prestashop.com/display/PS16/How+to+write+a+commit+message)!

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8179)

### CODE REVIEW

**Review by christianverardi — APPROVED** [review]

File: src/Adapter/Category/CategoryDataProvider.php [hunk 1]
```diff
@@ -150,20 +150,21 @@ public function getCategoriesByProduct(ObjectModel $product)
...
-                'id' => $category['id_category'],
-                'name' => $category['name'],
-                'breadcrumb' => $this->getBreadCrumb($category['id_category'])
-            ];
-        }
-
-        return $results;
+        $context = \ContextCore::getContext();
+        
+        return \DbCore::getInstance()->executeS('SELECT cl.`id_category` as "id", cl.`name`, (SELECT GROUP_CONCAT(cl.name ORDER BY c.`nleft` SEPARATOR "'.$delimiter.'")
```
Comment (LittleBigDev) [inline src/Adapter/Category/CategoryDataProvider.php:157]: Adapters are not model objects. They cannot build nor execute SQL queries. Please move this code in a model class.
For instance, previous version of this code was calling `Category::getAllCategoriesName()`. You need to provide a similar implementation.

File: src/Adapter/Category/CategoryDataProvider.php [hunk 2]
```diff
@@ -150,20 +150,21 @@ public function getCategoriesByProduct(ObjectModel $product)
...
+                                FROM `'._DB_PREFIX_.'category` c
+                                INNER JOIN `'._DB_PREFIX_.'category_shop` category_shop ON (category_shop.id_category = c.id_category AND category_shop.id_shop = '.$context->shop->id.')
+                                INNER JOIN `'._DB_PREFIX_.'category_lang` cl ON (c.`id_category` = cl.`id_category` AND cl.id_shop = ' . $context->shop->id . ' AND cl.id_lang = ' . $context->language->id . ')
+                                RIGHT JOIN `'._DB_PREFIX_.'category` c2 ON  c.`nleft` <= c2.`nleft` AND c.`nright` >= c2.`nright`
+                                WHERE 1 AND c.id_category IS NOT NULL AND c.`level_depth` > 0 AND c2.`id_category` = X.`id_category`) AS "breadcrumb"
```
Comment (LittleBigDev) [inline src/Adapter/Category/CategoryDataProvider.php:162]: Is there a specific reason for using `WHERE 1 AND` ?

**Review by LittleBigDev — CHANGES_REQUESTED** [review]
Hello,
Thanks for this improvement !
Though, some changes need to be done :

File: src/Adapter/Category/CategoryDataProvider.php (hunk 1 above)
Comment (AntoInta) [reply src/Adapter/Category/CategoryDataProvider.php:157]: Hi, may thank for your reply. 
I'll try to analyze and propose an extension of Category::getAllCategoriesName() if we want to retrieve the path of the category, in order to clean the Adapter.

Comment (LittleBigDev) [reply src/Adapter/Category/CategoryDataProvider.php:157]: Yeah, sorry about the delay. I know this makes it harder to fix when it's cold and old in your memories...

**Comment by marionf** [comment]
Changes are requested, so I remove the Waiting for QA label for now.
Please add it when the code review will be ok

**Comment by PierreRambaud** [comment]
Hi @AntoInta ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Pull request closed without merging**